ml_tools/feature_generator.py

```diff
@@ -0,0 +1,362 @@
+            for idx in ts_indices:
+                for index, time in data.iloc[:, idx].iteritems():
+                    data.iloc[:, idx][index] = time.hour
+            return data
+        else:
+            try:
+                for index, time in data[column].iteritems():
+                    data[column][index] = time.hour
+            except AttributeError:
+                data[column] = pd.to_datetime(data[column])
+                for index, time in data[
+                    column].iteritems():  # No idea why i need iteritems here but not in the other loop
+                    data[column][index] = time.hour
+            return data
+
+
+class DateTimeInfo(FeatureGenerator):
+    """Split timestamps into multiple features"""
+
+    @property
+    def name(self):
+        return 'datetime_info'
+
+    @property
+    def feature_type(self):
+        return 'generation'
+
+    @classmethod
+    def generate_feature(cls, data, column=None, **kwargs):
+        """Splits a timestamp into year, month, day, weekday and time of day
+
+        :param data: dataframe
+        :param column: column with timestamps to be split
+        :param kwargs: ignored
+        :return: dataframe with split column dropped and new columns added
+        """
+        if column is None:
+            raise ValueError('timestamp column must be given')
+        try:
+            data[f'{column}_year'] = [d.year for d in data[column]]
+            data[f'{column}_month'] = [d.month for d in data[column]]
+            data[f'{column}_day'] = [d.day for d in data[column]]
+            data[f'{column}_weekday'] = [d.weekday() for d in data[column]]
+            data[f'{column}_time'] = data[column].dt.time
+            data = data.drop([column], axis=1)
+        except AttributeError:
+            data[column] = pd.to_datetime(data[column])
+            data[f'{column}_year'] = [d.year for d in data[column]]
+            data[f'{column}_month'] = [d.month for d in data[column]]
+            data[f'{column}_day'] = [d.day for d in data[column]]
+            data[f'{column}_weekday'] = [d.weekday() for d in data[column]]
+            data[f'{column}_time'] = data[column].dt.time
+            data = data.drop([column], axis=1)
+        return data
+
+
+class ZipCodeInfo(FeatureGenerator):
+    """Generates new features from a zipcode
+
+    Searches zip codes and creates columns for state, county, city, latitude, longitude and time zone
+    """
+
+    @property
+    def name(self):
+        return 'zipcode_info'
+
+    @property
+    def feature_type(self):
+        return 'generation'
+
+    @classmethod
+    def generate_feature(cls, data, column=None, **kwargs):
+        """
+
+        :param data: dataframe containing zip codes
+        :param column: column label containing zip codes
+        :param kwargs: ignored
+        :return: dataframe with new columns for state, county, city, latitude, longitude and time zone
```

Comment:
Same thing about :type <blah>: :rtype: etc.  Same thing about the **kwargs ignored.